### PR TITLE
Martin/persist restore cache

### LINKED ISSUE #15846: Add buildkite functionality to persist and restore lagrange basis cache
We are storing the results of the lagrange basis computations when running dev unit tests (see #15802). Currently the tests lazily populate the cache, so that the first time you need the basis you compute and every other time you load from disk. We lose the disk at the end of the run,.

We would like to investigate persisting and restoring this cache for the dev unit tests since these basically never change. Ideally the flow would be something like

- restore the cache to proper directory via s3, nfs or whatever is easy 
- run tests
[ IDEALLY ]
- if the cache has changed (this can happen when new tests are added), push to store in smart way

NOTE: The total cache size is currently 8kb and not expected to grow over time.

### PR DESCRIPTION
**Pull request by martyall**
fixes #15846

### CODE REVIEW

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-buid-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by georgeee** [comment]
Instead of forming a tarfile with all the cache entries, I'd suggest to have each cache entry as a standalone `.gz` file (or just as is, without any packaging/compression).

My reasoning is that some of our tests are non-deterministic and they'll use different cache entries. If we overwrite an archive of one run with an archive of another run, we may have some entries always missing, whereas if we don't pack the cache entry, they will add up to a complete coverage eventually

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

File: buildkite/scripts/lagrange-cache-manager.sh [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+# Function to restore cache, preserving the directory structure
+restore_cache() {
+  local bucket_name=$1
+  local cache_dir=$2
+
+  echo "Restoring cache from bucket: gs://$bucket_name"
+  
+  if gsutil cp -r "gs://$bucket_name/pallas/*" "$cache_dir/pallas/" && \
+     gsutil cp -r "gs://$bucket_name/vesta/*" "$cache_dir/vesta/"; then
+    echo "Cache restored from GCS."
```
Comment (georgeee) [inline buildkite/scripts/lagrange-cache-manager.sh:19]:
```suggestion
    echo "Lagrange basis cache restored from GCS."
```

File: buildkite/scripts/lagrange-cache-manager.sh [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  else
+    echo "No cache found. Starting with a fresh cache."
```
Comment (georgeee) [inline buildkite/scripts/lagrange-cache-manager.sh:21]:
```suggestion
    echo "No lagrange basis cache found. Starting with a fresh cache."
```

File: buildkite/scripts/lagrange-cache-manager.sh [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  bucket_files=$(list_gcs_files "$bucket_name" "pallas")
+  bucket_files+=$'\n'
+  bucket_files+=$(list_gcs_files "$bucket_name" "vesta")
+
+  # Create a list of files currently in the cache directory, restricted to pallas and vesta directories
+  local_files=$(find "$cache_dir/pallas" "$cache_dir/vesta" -type f | sed 's#^'$cache_dir'/##')
+
+  echo "Local cache files: $local_files"
+
+  # Find new or changed files by comparing the lists
```
Comment (georgeee) [inline buildkite/scripts/lagrange-cache-manager.sh:47]: I suggest just doing `gsutil cp` with an `-n` flag.

This way `gsutil` will copy all files, but skip copying files that already exist on the bucket.

File: buildkite/scripts/lagrange-cache-manager.sh [hunk 4]
```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
+set -eo pipefail
+
+# Function to create cache directory with subdirectories for pallas and vesta
+create_cache_dir() {
+  local cache_dir=$1
+  mkdir -p "$cache_dir/pallas" "$cache_dir/vesta"
```
Comment (georgeee) [inline buildkite/scripts/lagrange-cache-manager.sh:7]: Suggest moving this line to `restore_cache` function (then `create_cache_dir` can be rmeoved)

**Review by georgeee — CHANGES_REQUESTED** [review]

**Comment by martyall** [comment]
@georgeee thanks for taking a look. But right now this is still in draft mode and I'm actually trying to determine whether this is worth doing at all (my theory is that it's not worth adding code to get negligible gain). It would be helpful to hold off looking at this until I request review.

**Comment by martyall** [comment]
I am going to close this as right now there is only marginal gain for added complexity. I will keep this branch alive in case this comes up in the near future

**Pull request closed without merging**